### ISSUE TITLE
Question on lazy loading related to a minor mode

### ISSUE THREAD

**Issue by maxwyb** [jwiegley/use-package#488]
I am new to Elisp and thanks for your time. I am adding a minor mode called `tern-mode` for JavaScript editing enhancements. It requires package `auto-complete` so I want to ensure it's there whenever `tern-mode` is loaded. Also, I mainly use `company-mode` as my autocompletion interface so I also want to add `company-tern` to my backends list when `tern-mode` is enabled.

I have this code now but the `use-package` block is never executed even though I opened a Javascript file and `tern-mode` is automatically loaded and enabled. (`tern-mode` is provided as a git repo here)

```lisp
(add-to-list 'load-path "~/.emacs.d/tern/emacs")
(autoload 'tern-mode "tern.el" nil t)
(add-hook 'js-mode-hook (lambda () (tern-mode t)))
(use-package auto-complete
  :ensure t
  :commands tern-mode
  :config
  ;; lines below are never executed
  (use-package company-tern
    :ensure t
    :config
    (add-to-list 'company-backends 'company-tern)))
```

I am doing almost the same for `irony-mode` and `company-irony` for C/C++, and this block works as expected. (`company-irony` is added to the `company-backends` list whenever a .c file is opened)

```lisp
(add-hook 'c-mode-hook 'irony-mode)
(use-package irony
  :ensure t
  :commands irony-mode
  :config
  (use-package company-irony
    :ensure t
    :config
    (add-to-list 'company-backends 'company-irony)))
```

So can someone give me a general direction on how to debug this problem? Greatly appreciate your help!

**Comment by maxwyb** [comment]
Alright I figured it out. **It's a bad idea for both `use-package` and lower-level Elisp `autoload` to handle lazy loading.** `use-package` can handle offline modules as well, so I just let it handle all packages by `:load-path` attribute. 

Also there is a hidden bug above: `auto-complete` package must be loaded *before* `tern-mode` begins to load, but they have the same level of precedence in the above code.

Here is the new block:
```elisp
(use-package auto-complete
  :ensure t)
(add-hook 'js-mode-hook (lambda () (tern-mode t)))
(use-package tern
  :load-path "tern/emacs"
  :commands tern-mode
  :config
  (use-package company-tern
    :ensure t
    :config
    (add-to-list 'company-backends 'company-tern)))
```

Thanks everyone for the help! Any suggestions to make the code more beautiful are welcomed.

**Comment by npostavs** [comment]
Assuming `tern-mode` is defined with `define-minor-mode` or equivalent, the `add-hook` could be just `(add-hook 'js-mode-hook #'tern-mode)`

**Comment by basil-conto** [comment]
> It's a bad idea for both `use-package` and lower-level Elisp `autoload` to handle lazy loading.

FWIW, there's no inherent conflict between `use-package` and `autoload` - they can be mixed and matched according to taste. In fact, certain forms of the former expand to calls to the latter to save typing on the user's part. Care to elaborate on why you think it's a bad idea to mix them?

**Comment by maxwyb** [comment]
Thanks for clarifying @basil-conto , and sorry that my conclusion is probably wrong... But the fact is that things begin to work correctly when I replace the `autoload` with `use-package :load-path`. Could you see my codes above and explain the reasons behind this, or am I getting something wrong? Thanks!

Note: I followed @npostavs 's suggestion and the result is the same...

**Comment by basil-conto** [comment]
@maxwyb 

> things begin to work correctly when I replace the `autoload` with `use-package :load-path`

These are two orthogonal concepts. `autoload`, which is what `use-package` uses behind the scenes when expanding `:commands`, `:mode`, etc., says "if function X is undefined when it is first called, then its definition can and should immediately be found in a file called Y". See [`(eintr) Autoload`](https://www.gnu.org/software/emacs/manual/html_node/eintr/Autoload.html) for more information on this.

`:load-path`, on the other hand, is just a shorthand for modifying the `load-path` variable at both byte-compile- and run-time, and says "when looking for *any* file to load, first look in directory X". As such, `:load-path` is only loosely tied to its enclosing `use-package` form (only its order of appearance w.r.t. other packages matters for side-effects). See [`(eintr) Loading Files`](https://www.gnu.org/software/emacs/manual/html_node/eintr/Loading-Files.html) for more on this.

In other words, it's not that `autoload` is unsuitable or incompatible with other things in your case; it's that you need to set the `load-path` appropriately for any (auto)loading to work.
